server/service/src/invoice/prescription/update/mod.rs

```diff
@@ -0,0 +1,450 @@
+use repository::{
+    Invoice, InvoiceLine, InvoiceLineRowRepository, InvoiceRowRepository, InvoiceRowStatus,
+    RepositoryError, StockLineRowRepository,
+};
+
+use crate::{
+    activity_log::{activity_log_entry, log_type_from_invoice_status},
+    invoice::query::get_invoice,
+    service_provider::ServiceContext,
+    WithDBError,
+};
+
+mod generate;
+mod validate;
+use generate::generate;
+use validate::validate;
+
+use self::generate::GenerateResult;
+
+#[derive(Clone, Debug, PartialEq)]
+pub enum UpdatePrescriptionStatus {
+    Picked,
+    Verified,
+}
+
+#[derive(Clone, Debug, PartialEq, Default)]
+pub struct UpdatePrescription {
+    pub id: String,
+    pub status: Option<UpdatePrescriptionStatus>,
+    pub patient_id: Option<String>,
+    pub clinician_id: Option<String>,
+    pub comment: Option<String>,
+    pub colour: Option<String>,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum UpdatePrescriptionError {
+    InvoiceDoesNotExist,
+    InvoiceIsNotEditable,
+    NotAPrescription,
+    NotThisStoreInvoice,
+    ClinicianDoesNotExist,
+    CanOnlyChangeToPickedWhenNoUnallocatedLines(Vec<InvoiceLine>),
+    // Name validation
+    OtherPartyDoesNotExist,
+    OtherPartyNotVisible,
+    OtherPartyNotAPatient,
+    // Internal
+    UpdatedInvoiceDoesNotExist,
+    DatabaseError(RepositoryError),
+    /// Holds the id of the invalid invoice line
+    InvoiceLineHasNoStockLine(String),
+}
+
+type OutError = UpdatePrescriptionError;
+
+pub fn update_prescription(
+    ctx: &ServiceContext,
+    patch: UpdatePrescription,
+) -> Result<Invoice, OutError> {
+    let invoice = ctx
+        .connection
+        .transaction_sync(|connection| {
+            let (invoice, status_changed) = validate(connection, &ctx.store_id, &patch)?;
+            let GenerateResult {
+                batches_to_update,
+                update_invoice,
+                unallocated_lines_to_trim,
+            } = generate(invoice, patch.clone(), connection)?;
+
+            InvoiceRowRepository::new(connection).upsert_one(&update_invoice)?;
+            let invoice_line_repo = InvoiceLineRowRepository::new(connection);
+
+            if let Some(stock_lines) = batches_to_update {
+                let repository = StockLineRowRepository::new(connection);
+                for stock_line in stock_lines {
+                    repository.upsert_one(&stock_line)?;
+                }
+            }
+
+            if let Some(lines) = unallocated_lines_to_trim {
```

Comment:
As per other comment, I don't think we will have unallocated lines for prescriptions